### ISSUE TITLE
```p{hr} = p{hr}``` "cannot prove goal by reflexivity"

### ISSUE THREAD

**Issue by oskgo** [EasyCrypt/easycrypt#136]
I found two cases where two variables look the same but cannot be proven to be equal. The first is because one variable is a program variable, while the other is a variable in easycrypt. For the second I have no idea. Looks like a bug to me at least.
```
require import AllCore.
type param.
module type A = {
  proc main(p: param) : bool
}.
section.
  declare module L <: A.

  local module Aux = {
    var r : bool
    proc main (p: param) = {
      r <@ L.main(p);
      return r;
}}.

lemma test &m p pr: Pr[Aux.main(p) @ &m : res] = pr.
proof.
byphoare (_: (glob L) = (glob L){m} /\ arg = p ==> res) => //.
proc.
seq 0 : (arg = p) (1%r) pr  _ 0%r (glob L = (glob L){m}); 1:by auto.
(* 1st case: p = p does not get simplified away here *)
simplify.
(* This is where the fun begins *)
auto. progress. reflexivity.
```

**Comment by fdupress** [comment]
A quick note that the severe overloading of `p` here is not at fault. The following fails in the same way.

```
require import AllCore.
type param.
module type A = {
  proc main(p: param) : bool
}.
section.
  declare module L <: A.

  local module Aux = {
    var r : bool
    proc main (p': param) = {
      r <@ L.main(p');
      return r;
}}.

lemma test &m p0 pr: Pr[Aux.main(p0) @ &m : res] = pr.
proof.
byphoare (_: (glob L) = (glob L){m} /\ arg = p0 ==> res) => //.
proc.
exists * arg, (glob L); elim * => [] p1 gL.
seq 0 : (p1 = p0) (1%r) pr _ 0%r (glob L = gL); 1:by auto.
auto.
move=> &hr [#] ->> _ _ <-. reflexivity.
```

**Comment by oskgo** [comment]
@fdupress Yeah. Maybe I should have filed a separate issue for that one. I have a hunch that there's some overloading going on in the second case as well though, but just harder to see or resulting from some bug.

**Comment by oskgo** [comment]
Issue #122 is related to the first example.

**Comment by mortensol** [comment]
Note that replacing ``arg = p`` with ``p = p`` seems to solve the problem in the sense that you no longer need to prove that ``p{hr} = p{hr}``.

**Comment by fdupress** [comment]
@mortensol, well yes, and so does removing the cut entirely.

But my feeling is indeed that `arg` is somehow being interpreted in a memory that's different from the one we expect, but that is somehow given the same name as the one we expect.
`arg` only really "exists" in the initial memory. But here, we use it in some intermediate memory. (Which happens to be the same, since it's the one obtained by executing `skip` on the initial memory, but that is irrelevant.) We *should* really reject its use in the `seq` event. (So in this case, I think we have that `&hr` is used as a name for both the initial memory within the `phoare` judgment, and for the memory at the program point we apply `seq` to.)

I haven't delved too deeply in the code yet, so can't confirm that this is what happens or how to fix it, but I think even that would be a symptom of something that's biting us in several places: the internal representation of program logic statements uses magical named memories with some hardcoded names for the contracts, instead of having them be reinterpretable in any memory that has the right "slots" for local variables. (As in, it literally treats tags `{1}` and `{2}` as memories, as opposed to treating them as names.)
I don't know if this is avoidable, but I know it caused me a bit of headache recently in trying to write a wrapper tactic around `transitivity`.

**Comment by oskgo** [comment]
I can no longer reproduce this. The original example gets a complaint about `arg` being unknown and replacing `arg` with `p{!hr}` works as expected.
